guide_drafts/extending-diesel.md

```diff
@@ -0,0 +1,385 @@
+```
+
+Diesel provides a proof-of-concept crate showing how to add new SQL types called
+`diesel_full_text_search`.
+These are the operators as they are defined in that crate.
+You'll notice all of the operators specify the backend,
+and many of them specify the return type.
+
+```
+diesel_infix_operator!(Matches, " @@ ", backend: Pg);
+diesel_infix_operator!(Concat, " || ", TsVector, backend: Pg);
+diesel_infix_operator!(And, " && ", TsQuery, backend: Pg);
+diesel_infix_operator!(Or, " || ", TsQuery, backend: Pg);
+diesel_infix_operator!(Contains, " @> ", backend: Pg);
+diesel_infix_operator!(ContainedBy, " <@ ", backend: Pg);
+```
+
+However, just declaring the operator by itself isn't very useful.
+This creates the types required by Diesel's query builder,
+but doesn't provide anything to help use the operator in real code.
+The structs created by these macros will have a `new` method,
+but that's not typically how you work with Diesel's query builder.
+
+- Infix operators are usually methods on the left hand side.
+- Postfix operators are usually methods on the expression.
+- Prefix operators are usually bare functions.
+
+For operators that you create with methods,
+you would typically create a trait for this.
+For example, here's how the `.eq` method gets defined by Diesel.
+
+```
+pub trait ExpressionMethods: Expression + Sized {
+    fn eq<T: AsExpression<Self::SqlType>>(self, rhs: T) -> Eq<Self, T::Expression> {
+        Eq::new(self, other.as_expression())
+    }
+}
+
+impl<T: Expression> ExpressionMethods for T {}
+```
+
+It's important to note that these methods are where you should put any type
+constraints.
+The structs defined by `diesel_*_operator!` don't know or care about what the
+types of the arguments should be.
+The `=` operator requires that both sides be of the same type,
+so we represent that in the type of `ExpressionMethods::eq`.
+
+You'll also notice that our argument is
+`AsExpression<Self::SqlType>`,
+not `Expression<SqlType = Self::SqlType>`.
+This allows Rust values to be passed as well as Diesel expressions.
+For example, we can do `text_col.eq(other_text_col)`,
+or `text_col.eq("Some Rust string")`.
+
+If the operator is specific to only one SQL type,
+we can represent that in our trait.
+
+```
+pub trait BoolExpressionMethods: Expression<SqlType = Bool> + Sized {
+    fn and<T: AsExpression<Bool>>(self, other: T) -> And<Self, T::Expression> {
+        And::new(self, other.as_expression())
+    }
+}
+
+impl<T: Expression<SqlType = Bool>> BoolExpressionMethods for T {}
+```
+
+Prefix operators are usually defined as bare functions.
+The code is very similar, but without the trait.
+Here's now `not` is defined in Diesel.
+
+```
+pub fn not<T: AsExpression<Bool>>(expr: T) -> Not<Grouped<T::Expression>> {
+    super::operators::Not::new(Grouped(expr.as_expression()))
+}
+```
+
+In this case we're using `Grouped` to add parenthesis around our argument.
```

Comment:
Maybe mention that this is `expression::grouped::Grouped`, which is currently only used internally (technically: hidden in the docs). Maybe we want to expose this in the future?